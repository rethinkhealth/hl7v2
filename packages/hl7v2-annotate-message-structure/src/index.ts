import type { Root, RootData } from "@rethinkhealth/hl7v2-ast";
import type { MessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import type { Plugin } from "unified";

/**
 * Unified plugin to infer HL7v2 message structure when missing.
 *
 * This plugin infers the message structure (MSH-9.3) from message code (MSH-9.1)
 * and trigger event (MSH-9.2) when the structure field is not present in the source.
 *
 * The inferred structure follows the convention: `{messageCode}_{triggerEvent}`
 *
 * @example
 * ```typescript
 * import { unified } from 'unified';
 * import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
 * import { hl7v2AnnotateMessage } from '@rethinkhealth/hl7v2-annotate-message';
 * import { hl7v2AnnotateMessageStructure } from '@rethinkhealth/hl7v2-annotate-message-structure';
 *
 * const processor = unified()
 *   .use(hl7v2Parser)
 *   .use(hl7v2AnnotateMessage)
 *   .use(hl7v2AnnotateMessageStructure);
 *
 * const tree = processor.parse(message);
 * await processor.run(tree);
 *
 * console.log(tree.data.messageInfo);
 * // { version: "2.5", messageCode: "ADT", triggerEvent: "A01", messageStructure: "ADT_A01" }
 * ```
 */
export const hl7v2AnnotateMessageStructure: Plugin<[], Root, Root> = () => {
  return (tree: Root) => {
    // Ensure data object exists
    if (!tree.data) {
      return tree;
    }

    // Get messageInfo from tree.data (cast to include messageInfo)
    const messageInfo = (tree.data as RootData & { messageInfo?: MessageInfo })
      .messageInfo;

    // If messageInfo doesn't exist, nothing to do
    if (!messageInfo) {
      return tree;
    }

    // If messageStructure is already set, don't override it
    if (messageInfo.messageStructure) {
      return tree;
    }

    // If we have both messageCode and triggerEvent, infer the structure
    if (messageInfo.messageCode && messageInfo.triggerEvent) {
      messageInfo.messageStructure = `${messageInfo.messageCode}_${messageInfo.triggerEvent}`;
    }

    return tree;
  };
};

export default hl7v2AnnotateMessageStructure;
