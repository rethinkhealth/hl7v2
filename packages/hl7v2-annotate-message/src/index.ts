import type { Root } from "@rethinkhealth/hl7v2-ast";
import type { MessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import { getMessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import type { Plugin } from "unified";

/**
 * Augment the RootData type to include messageInfo
 */
declare module "@rethinkhealth/hl7v2-ast" {
  // biome-ignore lint/nursery/useConsistentTypeDefinitions: module augmentation
  export interface RootData {
    /**
     * Message metadata extracted from MSH segment
     * Includes version (MSH-12), message code (MSH-9.1),
     * trigger event (MSH-9.2), and message structure (MSH-9.3)
     */
    messageInfo?: MessageInfo;
  }
}

/**
 * Unified plugin to annotate HL7v2 AST with message metadata.
 *
 * This plugin extracts message metadata from the MSH segment and stores it
 * in `Root.data.messageInfo` for downstream plugins to access.
 *
 * The metadata includes:
 * - `version`: MSH-12 (e.g., "2.5")
 * - `messageCode`: MSH-9.1 (e.g., "ADT")
 * - `triggerEvent`: MSH-9.2 (e.g., "A01")
 * - `messageStructure`: MSH-9.3 (e.g., "ADT_A01")
 *
 * @example
 * ```typescript
 * import { unified } from 'unified';
 * import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
 * import { hl7v2AnnotateMessage } from '@rethinkhealth/hl7v2-annotate-message';
 *
 * const processor = unified()
 *   .use(hl7v2Parser)
 *   .use(hl7v2AnnotateMessage);
 *
 * const tree = processor.parse(message);
 * await processor.run(tree);
 *
 * console.log(tree.data.messageInfo);
 * // { version: "2.5", messageCode: "ADT", triggerEvent: "A01", messageStructure: "ADT_A01" }
 * ```
 */
export const hl7v2AnnotateMessage: Plugin<[], Root, Root> = () => {
  return (tree: Root) => {
    const messageInfo = getMessageInfo(tree);

    // Initialize data object if it doesn't exist
    if (!tree.data) {
      tree.data = {};
    }

    // Store message info in Root.data
    tree.data.messageInfo = messageInfo;

    return tree;
  };
};

export default hl7v2AnnotateMessage;
