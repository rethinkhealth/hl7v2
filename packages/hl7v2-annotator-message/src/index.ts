/** biome-ignore-all lint/performance/noBarrelFile: fine */
export * from './plugin';
export type * from './types';
