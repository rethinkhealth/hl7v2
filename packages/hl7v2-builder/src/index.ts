/** biome-ignore-all lint/style/useUnifiedTypeSignatures: fine */
import type {
  Component,
  Field,
  FieldRepetition,
  Group,
  Root,
  RootContent,
  Segment,
} from "@rethinkhealth/hl7v2-ast";
import { u } from "unist-builder";

export function m(...children: RootContent[]): Root {
  return u("root", children);
}

export function g(name: string, ...children: Array<Segment | Group>): Group {
  const group: Group = u("group", children);
  group.name = name;
  return group;
}

// TODO: Add support for string[]
// export function s(name: string, ...fields: string[]): Segment;
// export function s(name: string, ...fields: string[] | Field[]): Segment {
export function s(name: string, ...fields: Field[]): Segment {
  const segment: Segment = u("segment", fields);
  segment.name = name;
  return segment;
}

type FieldValue = FieldRepetition | Component | string;
type Flattenable<T> = T | T[];

function flatten<T>(values: Flattenable<T>[]): T[] {
  return values.flatMap((value) => (Array.isArray(value) ? value : [value]));
}

function isFieldRepetition(value: FieldValue): value is FieldRepetition {
  return (
    typeof value === "object" &&
    value !== null &&
    value.type === "field-repetition"
  );
}

function isComponent(value: FieldValue): value is Component {
  return (
    typeof value === "object" && value !== null && value.type === "component"
  );
}

/** Empty field */
export function f(): Field;
export function f(...values: Flattenable<FieldValue>[]): Field;
export function f(...values: Flattenable<FieldValue>[]): Field {
  if (values.length === 0) {
    return u("field", [r()]);
  }

  const flat = flatten<FieldValue>(values);
  if (flat.length === 0) {
    return u("field", [r()]);
  }

  const repetitions: FieldRepetition[] = [];
  let pendingComponents: Array<Component | string> = [];

  const flushPending = () => {
    if (pendingComponents.length === 0) {
      return;
    }
    repetitions.push(r(...pendingComponents));
    pendingComponents = [];
  };

  for (const value of flat) {
    if (isFieldRepetition(value)) {
      flushPending();
      repetitions.push(value);
      continue;
    }

    if (isComponent(value) || typeof value === "string") {
      pendingComponents.push(value);
    }
  }

  flushPending();

  if (repetitions.length === 0) {
    return u("field", [r()]);
  }

  return u("field", repetitions);
}

export function r(): FieldRepetition;
export function r(
  ...components: Flattenable<Component | string>[]
): FieldRepetition;
export function r(
  ...components: Flattenable<Component | string>[]
): FieldRepetition {
  if (components.length === 0) {
    return u("field-repetition", [c()]);
  }

  const flat = flatten<Component | string>(components);
  if (flat.length === 0) {
    return u("field-repetition", [c()]);
  }

  return u(
    "field-repetition",
    flat.map((value) => (typeof value === "string" ? c(value) : value))
  );
}

export function c(): Component;
export function c(...values: Flattenable<string>[]): Component;
export function c(...values: Flattenable<string>[]): Component {
  if (values.length === 0) {
    return u("component", [u("subcomponent", "")]);
  }

  const flat = flatten<string>(values);
  if (flat.length === 0) {
    return u("component", [u("subcomponent", "")]);
  }

  return u(
    "component",
    flat.map((subcomponent) => u("subcomponent", subcomponent))
  );
}
