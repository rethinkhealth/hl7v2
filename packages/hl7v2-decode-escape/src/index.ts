import type { Root, Subcomponent } from '@rethinkhealth/hl7v2-ast';
import {
  DEFAULT_DELIMITERS,
  type HL7v2Delimiters,
} from '@rethinkhealth/hl7v2-utils';
import type { Plugin } from 'unified';
import { visit } from 'unist-util-visit';

export interface HL7v2DecodeOptions {
  delimiters?: Partial<HL7v2Delimiters>;
}

/**
 * Unified plugin to decode HL7v2 escape sequences in subcomponent literals.
 *
 * - Decodes \F\, \S\, \T\, \R\, \E\
 * - Decodes \Xdddd\ hex escapes
 * - Handles \.br\ line breaks
 * - Uses delimiters from Root.data.delimiters if available
 */
export const hl7v2DecodeEscapes: Plugin<[HL7v2DecodeOptions?], Root> = (
  options
) => {
  return (tree: Root) => {
    const delimiters =
      (tree.data as { delimiters?: Partial<HL7v2Delimiters> })?.delimiters ||
      options?.delimiters;

    visit(tree, 'subcomponent', (node: Subcomponent) => {
      const raw = node.value;
      node.value = decode(raw, {
        ...DEFAULT_DELIMITERS,
        ...delimiters,
      });
    });
  };
};

/**
 * Decode HL7v2 escape sequences according to HL7 v2.8 spec.
 *
 * @param value - The value to decode.
 * @param d - The delimiters to use.
 * @returns The decoded value.
 */
// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: this function must handle multiple HL7v2 escape cases and is as simple as possible given the requirements
function decode(value: string, d: typeof DEFAULT_DELIMITERS): string {
  if (!value?.includes(d.escape)) {
    return value;
  }

  let decoded = '';
  let i = 0;

  while (i < value.length) {
    if (value[i] === d.escape) {
      const end = value.indexOf(d.escape, i + 1);
      if (end === -1) {
        decoded += value.slice(i); // unterminated escape
        break;
      }

      const code = value.slice(i + 1, end);

      switch (code) {
        case 'F':
          decoded += d.field;
          break;
        case 'S':
          decoded += d.component;
          break;
        case 'R':
          decoded += d.repetition;
          break;
        case 'T':
          decoded += d.subcomponent;
          break;
        case 'E':
          decoded += d.escape;
          break;
        case '.br':
          decoded += '\r'; // or '\n' if you prefer LF
          break;
        case 'H':
        case 'N':
          // Highlight start/end: ignored for now
          break;
        default:
          if (code.startsWith('X') && code.length > 1) {
            decoded += decodeHexSequence(code.slice(1));
          } else {
            // Unknown escape: preserve as-is
            decoded += d.escape + code + d.escape;
          }
          break;
      }

      i = end + 1;
    } else {
      decoded += value[i++];
    }
  }

  return decoded;
}

/**
 * Decode HL7 \Xdddd\ hexadecimal escape sequences into characters.
 */
function decodeHexSequence(hex: string): string {
  let result = '';
  for (let i = 0; i < hex.length; i += 2) {
    const byte = hex.slice(i, i + 2);
    const codePoint = Number.parseInt(byte, 16);
    if (!Number.isNaN(codePoint)) {
      result += String.fromCharCode(codePoint);
    }
  }
  return result;
}
