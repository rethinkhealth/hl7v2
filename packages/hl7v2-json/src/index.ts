/** biome-ignore-all lint/suspicious/noExplicitAny: we should use a JSON type */
import type { HL7v2Node } from '@rethinkhealth/hl7v2-ast';
import type { Plugin } from 'unified';
import type { Node } from 'unist';

export const hl7v2Jsonify: Plugin<[], HL7v2Node, string> = function (): void {
  // biome-ignore lint/complexity/noUselessThisAlias: this is a plugin
  const self = this;

  function compiler(tree: Node): string {
    return JSON.stringify(toJson(tree as HL7v2Node), null, 2);
  }

  self.compiler = compiler;
};

/**
 * Convert an HL7v2Node tree into a simplified JSON representation.
 * Removes unist metadata and keeps only meaningful HL7v2 properties.
 */
export function toJson(node: HL7v2Node): any {
  const { type, name, index, value, delimiter, children } = node;

  const result: Record<string, any> = {
    type,
    ...(name ? { name } : {}),
    ...(index !== undefined ? { index } : {}),
    ...(value !== undefined ? { value } : {}),
    ...(delimiter ? { delimiter } : {}),
  };

  if (children && children.length > 0) {
    result.children = children.map(toJson);
  }

  return result;
}
