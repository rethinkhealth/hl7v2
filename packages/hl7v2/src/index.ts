import { hl7v2DecodeEscapes } from '@rethinkhealth/hl7v2-decode-escapes';
import { hl7v2Jsonify } from '@rethinkhealth/hl7v2-jsonify';
import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
import { unified } from 'unified';

/**
 * Create a new unified processor that already uses `hl7v2-parser`.
 */
export const parseHL7v2 = unified()
  .use(hl7v2Parser)
  .use(hl7v2DecodeEscapes)
  .use(hl7v2Jsonify)
  .freeze();
