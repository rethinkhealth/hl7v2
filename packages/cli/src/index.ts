#!/usr/bin/env node
/** biome-ignore-all lint/suspicious/noConsole: This is a CLI */
import { readFileSync } from 'node:fs';
import { createInterface } from 'node:readline';

import { type HL7v2Node, parseHL7 } from '@rethinkhealth/hl7v2';
import { Command } from 'commander';

const program = new Command();

program.name('hl7').description('HL7v2 CLI Tool').version('0.1.0');

program
  .command('parse')
  .description('Parse an HL7v2 message')
  .argument('[file]', 'Path to HL7 file')
  .action(async (file) => {
    let message: string;

    if (file) {
      message = readFileSync(file, 'utf8');
    } else {
      // Interactive mode
      const rl = createInterface({
        input: process.stdin,
        output: process.stdout,
      });
      message = await new Promise<string>((resolve) =>
        rl.question('Paste HL7 message:\n', resolve)
      );
      rl.close();
    }

    const dom = parseHL7(message);

    // Show parsed tree
    console.log('\nParsed HL7 DOM:\n');
    console.log(JSON.stringify(dom, null, 2));

    // Basic stats
    const segments = dom.children?.length || 0;
    console.log('\nStats:');
    console.log(`- Segments: ${segments}`);
    console.log(`- Characters: ${message.length}`);
    console.log(`- Message Type: ${getMessageType(dom)}`);
  });

program.parse(process.argv);

// Extract MSH-9 for message type
function getMessageType(dom: HL7v2Node): string {
  const msh = dom.children?.find((n: HL7v2Node) => n.name === 'MSH');
  return msh?.children?.[8]?.value || 'Unknown';
}
