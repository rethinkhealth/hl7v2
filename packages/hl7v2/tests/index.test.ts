import { describe, expect, it } from 'vitest';
import { parseHL7v2 } from '../src';

describe('parseHL7v2 (holistic)', () => {
  it('parses MSH + PID and decodes escapes, then stringifies to JSON', async () => {
    const msg = [
      // MSH with delimiters and sender
      'MSH|^~\\&|SENDER|FAC|RCVR|FAC|20250101010101||ADT^A01|MSG00001|P|2.5',
      // PID with components, repetitions and escapes (\S\ for ^)
      'PID|1||12345^^^HOSP^MR||DOE\\S\\JOHN^A&J|19700101|M',
    ].join('\r');

    const file = await parseHL7v2.process(msg);

    expect(file.value).toMatchSnapshot();
  });

  it('handles trailing delimiters and preserves empty structures appropriately', async () => {
    const msg = 'PID|A^B&|C~|';
    const file = await parseHL7v2.process(msg);

    expect(file.value).toMatchSnapshot();
  });

  it('handles decoding escapes', async () => {
    const msg = 'PID|1|Value\\F\\More\\S\\Stuff\\R\\Again\\T\\Sub\\E\\Escaped|';
    const file = await parseHL7v2.process(msg);

    expect(file.value).toMatchSnapshot();
  });

  it('parses multiple segments with MSH as the first segment', async () => {
    const msg = [
      // MSH
      'MSH|^~\\&|SENDER|FAC|RCVR|FAC|20250101010101||ADT^A01|MSG00001|P|2.5',
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R',
    ].join('\r');

    const file = await parseHL7v2.process(msg);
    expect(file.value).toMatchSnapshot();
  });

  it('parses multiple segments without MSH as the first segment', async () => {
    const msg = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R',
    ].join('\r');

    const file = await parseHL7v2.process(msg);

    expect(file.value).toMatchSnapshot();
  });

  it('parses correctly similar messages with different carriage returns', async () => {
    const msg_with_cr = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R',
    ].join('\r');

    const msg_with_lf = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R',
    ].join('\n');

    const file_cr = await parseHL7v2.process(msg_with_cr);
    const file_lf = await parseHL7v2.process(msg_with_lf);

    expect(file_cr.value).toEqual(file_lf.value);
  });

  it('parses correctly similar messages with or without trailing field separators', async () => {
    const msg_with_trailing_field_separator = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC|',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R|',
    ].join('\r');

    const msg_without_trailing_field_separator = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R',
    ].join('\r');

    const file_cr = await parseHL7v2.process(msg_with_trailing_field_separator);
    const file_lf = await parseHL7v2.process(
      msg_without_trailing_field_separator
    );

    expect(file_cr.value).toEqual(file_lf.value);
  });

  it('parses correctly similar messages with or without trailing field separators and empty last field', async () => {
    const msg_with_trailing_field_separator = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC|',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^|',
    ].join('\r');

    const msg_without_trailing_field_separator = [
      // PID
      'PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC',
      // OBX
      'OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^',
    ].join('\r');

    const file_cr = await parseHL7v2.process(msg_with_trailing_field_separator);
    const file_lf = await parseHL7v2.process(
      msg_without_trailing_field_separator
    );

    expect(file_cr.value).toEqual(file_lf.value);
  });
});
