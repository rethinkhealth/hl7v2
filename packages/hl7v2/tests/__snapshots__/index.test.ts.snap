// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseHL7v2 (holistic) > handles decoding escapes 1`] = `"PID|1|Value\\F\\More\\S\\Stuff\\R\\Again\\T\\Sub\\E\\Escaped|"`;

exports[`parseHL7v2 (holistic) > handles trailing delimiters and preserves empty structures appropriately 1`] = `"PID|A^B&|C~|"`;

exports[`parseHL7v2 (holistic) > parses MSH + PID and decodes escapes, then stringifies to JSON 1`] = `
"MSH|^~\\&|SENDER|FAC|RCVR|FAC|20250101010101||ADT^A01|MSG00001|P|2.5
PID|1||12345^^^HOSP^MR||DOE\\S\\JOHN^A&J|19700101|M"
`;

exports[`parseHL7v2 (holistic) > parses multiple segments with MSH as the first segment 1`] = `
"MSH|^~\\&|SENDER|FAC|RCVR|FAC|20250101010101||ADT^A01|MSG00001|P|2.5
PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC
OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R"
`;

exports[`parseHL7v2 (holistic) > parses multiple segments without MSH as the first segment 1`] = `
"PID|||PATID1234^5^M11^ADT1^MR^UNIVERSITY HOSPITAL~123_456_789^^^USSSA^SS||EVERYMAN^ADAM^A^III||19_610_615|M||C|1200 N ELM STREET^^GREENSBORO^NC^27_401-1020|GL|(919)379-1212|(919)271-3434||S||PATID12345001^2^M10^ADT1^AN^A|123_456_789|9-87_654^NC
OBR|1|845439^GHH OE|1045813^GHH LAB|1554-5^GLUCOSE^LN|||200202150730|||||||||DOCT^KILDARE^JAMES^A|||||||200202150930||F|||^^^^^R"
`;

exports[`parseHL7v2 (holistic) > parses multiple segments without MSH as the first segment 2`] = `
[
  {
    "fields": [
      "",
      "",
      [
        [
          "PATID1234",
          "5",
          "M11",
          "ADT1",
          "MR",
          "UNIVERSITY HOSPITAL",
        ],
        [
          "123_456_789",
          "",
          "",
          "USSSA",
          "SS",
        ],
      ],
      "",
      [
        "EVERYMAN",
        "ADAM",
        "A",
        "III",
      ],
      "",
      "19_610_615",
      "M",
      "",
      "C",
      [
        "1200 N ELM STREET",
        "",
        "GREENSBORO",
        "NC",
        "27_401-1020",
      ],
      "GL",
      "(919)379-1212",
      "(919)271-3434",
      "",
      "S",
      "",
      [
        "PATID12345001",
        "2",
        "M10",
        "ADT1",
        "AN",
        "A",
      ],
      "123_456_789",
      [
        "9-87_654",
        "NC",
      ],
    ],
    "segment": "PID",
  },
  {
    "fields": [
      "1",
      [
        "845439",
        "GHH OE",
      ],
      [
        "1045813",
        "GHH LAB",
      ],
      [
        "1554-5",
        "GLUCOSE",
        "LN",
      ],
      "",
      "",
      "200202150730",
      "",
      "",
      "",
      "",
      "",
      "",
      "",
      "",
      [
        "DOCT",
        "KILDARE",
        "JAMES",
        "A",
      ],
      "",
      "",
      "",
      "",
      "",
      "",
      "200202150930",
      "",
      "F",
      "",
      "",
      [
        "",
        "",
        "",
        "",
        "",
        "R",
      ],
    ],
    "segment": "OBR",
  },
]
`;
